### ISSUE TITLE
AttributeError: object has no attribute 'decode' for PAYLOAD_REGISTRY class

### ISSUE THREAD

**Issue by folt** [aio-libs/aiohttp#5281]
🐞 **Describe the bug**
While working with `PAYLOAD_REGISTRY` I came across a problem. After examining the information, I found an error. The fact is that the Response class implements a property that calls the [self._body.decode](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/web_response.py#L587) method expecting a string there. But when using `PAYLOAD_REGISTRY` (I used `JsonPayload`) in `self._body` there is an instance of the payload class where there is no implementation of the `decode` method.

💡 **To Reproduce**
```
from typing import Mapping, Optional
from types import MappingProxyType
from http import HTTPStatus

from aiohttp import web
from aiohttp import payload
from aiohttp import web_request
from aiohttp import web_middlewares
from aiohttp import PAYLOAD_REGISTRY
from aiohttp import web_exceptions


def format_http_error(http_error_cls, message: Optional[str] = None,
                      ) -> web_exceptions.HTTPException:
    status = HTTPStatus(http_error_cls.status_code)

    error = {
        'code': status.name.lower(),
        'message': message or status.description
    }

    return http_error_cls(body={'error': error})


async def handle():
    raise web_exceptions.HTTPNotFound()


@web_middlewares.middleware
async def error_middleware(request: web_request.Request, handler):
    try:
        return await handler(request)
    except web_exceptions.HTTPException as err:
        err = format_http_error(err.__class__, err.text)
        raise err


app = web.Application(
    middlewares=[error_middleware]
)

PAYLOAD_REGISTRY.register(payload.JsonPayload, (Mapping, MappingProxyType))

app.add_routes([web.get('/', handle)])

if __name__ == '__main__':
    web.run_app(app)
```

📋 **Logs/tracebacks**
```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/folt/Documents/Work/geodb/.venv/lib/python3.8/site-packages/aiohttp/web_protocol.py", line 422, in _handle_request
    resp = await self._request_handler(request)
  File "/Users/folt/Documents/Work/geodb/.venv/lib/python3.8/site-packages/aiohttp/web_app.py", line 499, in _handle
    resp = await handler(request)
  File "/Users/folt/Documents/Work/geodb/.venv/lib/python3.8/site-packages/aiohttp/web_middlewares.py", line 118, in impl
    return await handler(request)
  File "/Users/folt/Documents/Work/geodb/qq.py", line 35, in error_middleware
    raise err
aiohttp.web_exceptions.HTTPNotFound: Not Found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/folt/Documents/Work/geodb/.venv/lib/python3.8/site-packages/aiohttp/web_protocol.py", line 485, in start
    resp, reset = await task
  File "/Users/folt/Documents/Work/geodb/.venv/lib/python3.8/site-packages/aiohttp/web_protocol.py", line 427, in _handle_request
    status=exc.status, reason=exc.reason, text=exc.text, headers=exc.headers
  File "/Users/folt/Documents/Work/geodb/.venv/lib/python3.8/site-packages/aiohttp/web_response.py", line 650, in text
    return self._body.decode(self.charset or "utf-8")
AttributeError: 'JsonPayload' object has no attribute 'decode'
```

📋 **Your version of the Python**
`
Python 3.8.5
`

📋 **Your version of the aiohttp/yarl/multidict distributions**
$ python -m pip show aiohttp
```
Name: aiohttp
Version: 3.7.2
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: Nikolay Kim
Author-email: fafhrd91@gmail.com
License: Apache 2
Location: /Users/folt/Documents/Work/geodb/.venv/lib/python3.8/site-packages
Requires: async-timeout, multidict, typing-extensions, chardet, attrs, yarl
Required-by: aiohttp-jinja2, aiohttp-debugtoolbar, aiohttp-apispec
```

$ python -m pip show multidict
```
Name: multidict
Version: 5.0.0
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /Users/folt/Documents/Work/geodb/.venv/lib/python3.8/site-packages
Requires: 
Required-by: yarl, aiohttp
```

$ python -m pip show yarl
```
Name: yarl
Version: 1.6.2
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl/
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /Users/folt/Documents/Work/geodb/.venv/lib/python3.8/site-packages
Requires: idna, multidict
Required-by: aiohttp
```

Also found a related issue #2928 . I solved the error by implementing my JsonPayload.
I see 2 ways to solve this error.
1. Implement the `decode` method in the `JsonPayload` class
2. In the `def text(self)` method, check the type of `_body` and, based on this check, decide whether to call the `decode` method for `_body`.

I am ready to contribute, but I would like to discuss possible solutions.

**Comment by i3Cheese** [comment]
Can you do pull request?

**Comment by folt** [comment]
> Can you do pull request?

I made a pull request #5350 . But it was not considered.

**Comment by Dreamsorcerer** [comment]
> I made a pull request #5350 . But it was not considered.

You appear to have closed your own PR on the same day you opened it...

No one is going to review a PR the author has already closed.

**Comment by folt** [comment]
> > I made a pull request #5350 . But it was not considered.
> 
> You appear to have closed your own PR on the same day you opened it...
> 
> No one is going to review a PR the author has already closed.

I have prepared a new pull request

**Comment by Dreamsorcerer** [comment]
Sorry, I've just had a proper look at this. I think the cause is another unrelated regression. Your original code runs fine on aiohttp 3.6 and 3.8. It's just 3.7 that has this problem.

Also, if you were running with warnings enabled, you'd see that the `body` argument you are passing to `HTTPException` is deprecated, so your code will also fail on master.

**Comment by Dreamsorcerer** [comment]
If you can create a different reproducer that demonstrates a problem on master, then we can reopen this.
